src/parser/mod.rs

```diff
@@ -1013,175 +1189,22 @@ impl<'a> Parser<'a> {
-                    let expr = self.parse_subexpr(self.dialect.prec_value(Precedence::PlusMinus))?;
-                    Ok(Expr::Prior(Box::new(expr)))
-                }
-                Keyword::MAP if self.peek_token() == Token::LBrace && self.dialect.support_map_literal_syntax() => {
-                    self.parse_duckdb_map_literal()
-                }
-                // Here `w` is a word, check if it's a part of a multipart
-                // identifier, a function call, or a simple identifier:
-                _ => match self.peek_token().token {
-                    Token::LParen | Token::Period => {
-                        let mut id_parts: Vec<Ident> = vec![w.to_ident()];
-                        let mut ends_with_wildcard = false;
-                        while self.consume_token(&Token::Period) {
-                            let next_token = self.next_token();
-                            match next_token.token {
-                                Token::Word(w) => id_parts.push(w.to_ident()),
-                                Token::Mul => {
-                                    // Postgres explicitly allows funcnm(tablenm.*) and the
-                                    // function array_agg traverses this control flow
-                                    if dialect_of!(self is PostgreSqlDialect) {
-                                        ends_with_wildcard = true;
-                                        break;
-                                    } else {
-                                        return self
-                                            .expected("an identifier after '.'", next_token);
-                                    }
-                                }
-                                Token::SingleQuotedString(s) => {
-                                    id_parts.push(Ident::with_quote('\'', s))
-                                }
-                                _ => {
-                                    return self
-                                        .expected("an identifier or a '*' after '.'", next_token);
-                                }
-                            }
+            // We first try to parse the word as the prefix of an expression.
+            // For example, the word INTERVAL in: SELECT INTERVAL '7' DAY
+            Token::Word(w) => match self.try_parse(|parser| parser.parse_expr_by_keyword(&w)) {
+                Ok(expr) => Ok(expr),
+                // Word does not indicate the start of a complex expression, try to parse as identifier
+                Err(ParserError::BranchAbandoned) => Ok(self.parse_ident_expr(&w)?),
+                // Word indicates the start of a complex expression, try to parse as identifier if the
+                // dialect does not reserve it, otherwise return the original error
+                Err(e) => {
+                    if !self.dialect.is_reserved_for_identifier(w.keyword) {
```

Comment:
one thing I'm wondering regarding the original problem is whether the underlying issue is rather the keywords like interval not being properly gated to only the dialects that support them?
I suspect there are cases that could benefit from not having the parser mistaking keywords on dialects that don't support them though (like in this `self.dialect.is_reserved_for_identifier`) - but not super sure if there are cases in this scenario with exprs here that strictly require it

Reply:
I think the problem lies in dialects that interpret keywords differently in different contexts (I guess by looking ahead). For example, in Snowflake, the INTERVAL word can be a keyword for an INTERVAL expr or an ident:
```
create or replace temporary table test(interval int);
insert into test (interval) values (1), (2);
select current_date() + interval '1 day', max(interval) from test;
```

I see four possible cases for the parser to handle:
1. The word is part of a syntactically correct expression, e.g. `INTERVAL '1 day'` --> handled
2. The word is part of a syntactically incorrect expression, e.g. `INTERVAL '1 dai'` --> handled
3. The word is an identifier and the dialect does not reserve it, e.g. `SELECT MAX(interval) FROM tbl` --> not handled
4. The word is an identifier and the dialect reserves it, e.g. `SELECT MAX(from) FROM test;` --> not handled perfectly, but we are OK with the parser to be more accepting, as long as over time we give maintainers the tools to add reserved keywords as needed.

Reply:
ah got it! I was under the impression the issue was snowflake not supporting interval as a keyword, thanks for clarifying